### ISSUE TITLE
Goto definition works once, then fails

### ISSUE THREAD

**Issue by Stibbons** [DamnWidget/anaconda#367]
Hello

I manage to get autocomplete, autodoc and goto definition to work.
When I use "Goto doc" on one file, I go to the definition of an object A located in another module B, but when I go back on the first file, the import statement is changed in "can't import B"...

**Comment by Stibbons** [comment]
any update about my issue? "Goto" works once and seem to delete some data so it doesn't work any more until next sublime restart ... :(

**Comment by DamnWidget** [comment]
Couldn't reproduce this, it works fine for me

**Comment by Stibbons** [comment]
Some information about my setup:
- virtual env configured in project with 
```
		"anaconda_linting": true,
		"anaconda_linting_behaviour": "always",
		"extra_paths":
		[
			"/home/USER/projects/path/to/sandbox/lib/python2.7/site-packages/",
		],
		"python_interpreter": "/home/USER/projects/path/to/sandbox/lib/python2.7/bin/python2.7",
```

**Comment by DamnWidget** [comment]
I don't think you need the `extra_path`, anaconda will be able to see whatever your configured interpreter  can see with no additions, try to remove it from your configuration, restart ST3 and try again.

Some users have reported strange behavior with ST3 since one of their last releases, in some cases, users needed to delete the ST3 cache completely to make things works, I don't thinks that is the case in here but is something to take into account.

**Comment by divanovGH** [comment]
I have the same problem, and not use "extra_paths"! (((

**Comment by codespresso** [comment]
Happening for me as well

**Comment by DamnWidget** [comment]
Can you retrieve the documentation for the object that you want to jump more than once? Is there anything in the ST3 console logs? Anything in the JsonServer logs?

**Comment by codespresso** [comment]
No, not able to retrieve documentation also. Nothing in console logs, just says Unable to find "<object name>" in status bar. Anything specific I can look for in JsonServer logs?

I am using virtualenv and have mentioned the python_interpreter in the project settings.

**Comment by DamnWidget** [comment]
You can find logs depending on your operating system on:

* Linux: ~/.cache/jedi/<project_name>/anaconda_jsonserver.log
* OS X: ~/Library/Caches/Jedi/<project_name>/anaconda_jsonserver.log
* Windows: %APPDATA%\\Jedi\\Jedi\\<project_name>\\anaconda_jsonserver.log

**Comment by codespresso** [comment]
2016-05-12 18:17:23,942: INFO    : Incomming connection from ('127.0.0.1', 64249)
2016-05-12 18:17:26,429: INFO    : client requests: goto
2016-05-12 18:17:26,510: INFO    : About push back to ST3: {"goto": [["/Users/codespresso/Workspace/virtualenvs/cbort/lib/python2.7/site-packages/djangorestframework-3.3.1-py2.7.egg/rest_framework/generics.py", 25, 7]], "success": true, "uid": "565f684ed086429aa268aa25a71cc85f"}

2016-05-12 18:17:26,544: INFO    : client requests: lint
2016-05-12 18:17:26,894: INFO    : About push back to ST3: {"vid": 17, "errors": [], "uid": "6e16b9dc5347464bb512ffbe89e5b2c3", "success": true}

2016-05-12 18:17:26,894: INFO    : client requests: lint
2016-05-12 18:17:26,961: INFO    : About push back to ST3: {"vid": 17, "errors": [], "uid": "795e646a1dee4f9a9a4c8bff0c816661", "success": true}

2016-05-12 18:17:27,539: INFO    : client requests: lint
2016-05-12 18:17:27,690: INFO    : About push back to ST3: {"vid": 15, "errors": [{"level": "W", "raw_error": "[W] PEP 8 (E303): too many blank lines (4)", "lineno": 102, "offset": 0, "message": "[w] pep 8 (%s): %s", "underline_range": true}], "uid": "cb2fafa8ce794ea28ed4573ed6c881dd", "success": true}

2016-05-12 18:17:28,791: INFO    : client requests: goto
2016-05-12 18:17:28,821: INFO    : About push back to ST3: {"goto": [], "success": true, "uid": "064d301d14c44166aa936bba71580471"}

**Comment by DamnWidget** [comment]
What are you trying to do goto? Your own code in other files?

**Comment by codespresso** [comment]
Was trying to goto object in external library in python site-packages. Strangely it seems to be working fine after todays sublime text update.

Not sure if its the reinstall or something got fixed in the update.

-- 
Rohan Gupta


On 13 May 2016 at 3:25:06 PM, Oscar Campos (notifications@github.com) wrote:

What are you trying to do goto? Your own code in other files?

**Comment by DamnWidget** [comment]
Dunno, I couldn't replicate it ever :/

**Comment by aaakbar** [comment]
Seems to be working fine for the package/module parts of references but not for class/obj parts of references; the behavior is consistent regardless of the context (import statements or other parts of code behave the same).

**Scenario 1**
`from [package] import [module/class]`  (tested on: `from django.contrib.auth.modules import User`) 
* goto definition for `[package]` or any of its parts (tested with:  `django.contrib.auth.modules`, `django`, `contrib`, `contrib.auth`, `auth.modules`) works nicely every time it's triggered
* goto definition for `[module]` (tested with: `User`) works only the first time

**Scenario 2**
if importing module/package directly using a simple import statement (tested on: `import mongoengine`), go to definiton works each time

**Scenario 3**
if importing class/callable directly using a simple import statement (tested on: `import mongoengine.StringField()`)
* goto def. will work each time on package/module part (tested with: `mongodb`)
* will only work once for class part (tested with: `StringField`)

**Scenario 4**
when using package/module resources in your logic:
`[var] = [package/module].[obj/callable]()` (e.g.: `user_id = mongoengine.StringField()`)
* goto definition works each time on `[package/module]` (in example: `mongoengine`) part of reference,
* but works only the first time on the `[obj/callable]` (in example: `StringField`) part of the reference

**Comment by DamnWidget** [comment]
Thank you for quite a complete report @aaakbar, I wish all the reports where like this one, I have my plate full right now preparing the v2 release and fixing other bugs but I will investigate deeper on this one as soon as I have a gap.

**Comment by DamnWidget** [comment]
I got some time after lunch to test this again. I created a new virtual environment using Python3 were I installed django and mongoengine to replicate Scenarios 1 to 4.

```bash
$ pip freeze
Django==1.9.7
mongoengine==0.10.6
pymongo==3.2.2
```

**Scenario 1**
* I created a new django project using `django-admin startproject mysite`
* I edited opened the mysite directory with ST3
* I saved new project with project settings:
```json
{
	"build_systems":
	[
		{
			"file_regex": "^[ ]*File \"(...*?)\", line ([0-9]*)",
			"name": "Anaconda Python Builder",
			"selector": "source.python",
			"shell_cmd": "\"/home/damnwidget/.virtualenvs/django/bin/python\" -u \"$file\""
		}
	],
	"folders":
	[
		{
			"path": "."
		}
	],
	"settings":
	{
		"python_interpreter": "~/.virtualenvs/django/bin/python"
	}
}
```
* I did open `mysite/mysite/urls.py
* I added `from django.contrib.auth.models import User` and `User.check()` a couple of lines after the import

Results:
* goto definition for package level works every time
* goto definition for `User` in both the import line or the `User.check()` line works every time

**Scenario 2**
* I added new line `import mongoengine`

Results:
Goto definition for the package worked every time

**Scenario 3**
* I added new line `mongoengine.StringField()`

Results:
* goto definition in the package level worked every time
* goto definition in `StringField()` worked every time

**Scenario 4**
* I tested anaconda code itself that is the only Python project that I am currently maintaining/working on 

Results:
* goto definition in package level worked every time
* goto definition in class, methods, etc worked every time

## Test env. characteristics
* OS: Manjaro 16.06-pre1 Daniella
* ST3: Build 3114
* Anaconda Versions: v1.4.26 and v2.0.0-rc1
* Plugins: anaconda, anaconda_rust, anaconda_php, JSON Reindent, Origami, Project Manager, Package Control

I am not using any virtual machine or fancy configuration

Resolution: **Not replicable**

Probably environmental issue

**Comment by aaakbar** [comment]
@DamnWidget Thanks for looking into the problem, for the detailed report, as well as for the great dev tool Anaconda is. Reading as much as possible regarding this error I see that users with local project installations (everything on the same host) don't have the same problem at all.

My conclusion after reading your report is much the same as yours - the problem must be environment. The fact that GoTo Def. always works the first time (and works for module/package entities) still seems weird but I'll try delving into Anaconda implementation and check its nuts and bolts to clarify this for myself.

**Comment by DamnWidget** [comment]
@aaakbar if you don't want to spend time in something that probably is fixed in v2.0.0 (as all the code involving remote environment integration has been rewritten in v2.0.0-rc1) checkout the v2.0.0-rc1 branch and take a look at the new documentation for [vagrant](https://github.com/DamnWidget/anaconda/wiki/Anaconda-v2:-Using-a-remote-python-interpreter-through-Vagrant) environments in version 2, there is a flexible "remote worker" as well, take a look at the Anaconda v2 documentation in general in the wiki and give it a try.

A new feature of anaconda's v2 remote workers (included vagrant ones) is that you can map directories to fix or minimize the problems with Goto Definition and other related commands.

**Comment by DamnWidget** [comment]
Probably worth a look at #562
